pkg/controllers/provisioning/provisioner.go

```diff
@@ -0,0 +1,120 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provisioning
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/awslabs/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/awslabs/karpenter/pkg/cloudprovider"
+	"github.com/awslabs/karpenter/pkg/controllers/allocation"
+	"github.com/awslabs/karpenter/pkg/controllers/allocation/scheduling"
+	v1 "k8s.io/api/core/v1"
+	"knative.dev/pkg/logging"
+)
+
+var (
+	MaxBatchWindow = time.Second * 10
+	MinBatchWindow = time.Second * 1
+)
+
+// Provisioner is a stateful, threadsafe controller. Pods are enqueud and
+// batched, capacity is launched, and pods are bound to the new capacity.
+type Provisioner struct {
+	// State
+	*v1alpha5.Provisioner
+	instanceTypes []cloudprovider.InstanceType
+	pods          chan *v1.Pod
+	results       chan error
+	cancelFunc    context.CancelFunc
+
+	// Dependencies
+	scheduler *scheduling.Scheduler
+	launcher  *allocation.Launcher
+}
+
+// Start the provisioner's loop
+func (p *Provisioner) start(ctx context.Context) {
+	logging.FromContext(ctx).Info("Starting provisioner")
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			if err := p.provision(ctx); err != nil {
+				logging.FromContext(ctx).Errorf("Provisioning failed, %s", err.Error())
+			}
+		}
+	}
+}
+
+func (p *Provisioner) stop(ctx context.Context) {
+	logging.FromContext(ctx).Info("Stopping provisioner")
+	p.cancelFunc()
+	close(p.pods)
+	close(p.results)
+}
+
+func (p *Provisioner) provision(ctx context.Context) (err error) {
+	// Wait for a batch of pods
+	pods := p.Batch(ctx)
+	// Send results
+	defer func() {
+		for i := 0; i < len(pods); i++ {
+			p.results <- err
```

Comment:
Could you add a comment clarifying here? It's a bit hard to get my head around why we'd send the same `err` back along the channel for each pod. Also how are these errors "matched up" with the Pod(s) by whatever is pulling results off of this channel?